Initial review of ftdc importer

visualzhou opened the pull request:
Both "base" and "review" branches are created only for the review since "master" has a long history without a clean initial commit. All changes will be cherry-picked to master later.

visualzhou committed:
Initial commit: FTDC to Victoria Metrics Importer

visualzhou committed:
feat: add support for different metric types (Int32, Int64, Double) - Add MetricType enum and metric_type field to MetricValue - Implement type conversion for Int32 and Int64 to f64 - Add comprehensive tests for each metric type - Fix linting issues and improve code organization

visualzhou committed:
Implement FTDC file reading and streaming

visualzhou committed:
Implement Victoria Metrics integration, CLI application, and code improvements - Add Victoria Metrics client with InfluxDB Line Protocol support - Implement CLI application with import and check modes - Fix variable mutability warnings - Refactor code organization - Add comprehensive tests

visualzhou committed:
Fix FTDC decompression and implement column-oriented processing

visualzhou committed:
Update development notes to reflect Phase 2 progress

visualzhou committed:
More parsing.

visualzhou committed:
More decode.

visualzhou committed:
Add varint and its tests.

visualzhou committed:
Update gitignore.

visualzhou committed:
Format.

visualzhou committed:
Fix warnings.

visualzhou committed:
Merge tests.

visualzhou committed:
Split metrics array to its own file.

visualzhou committed:
Fix test_decode_metrics_array itself.

visualzhou committed:
Fix timestamp metrics count.

visualzhou committed:
Fix timestamp of document.

visualzhou committed:
Print out the metrics count earlier.

visualzhou committed:
Add bool.

visualzhou committed:
Support Decimal 128.

milkie committed:
better debug output formatting

milkie committed:
reformat, and remove erroneous ref doc parse from type 1

milkie committed:
more fix

milkie committed:
more fix

visualzhou committed:
Dedup varint.

visualzhou committed:
Dedup errors.

visualzhou committed:
Add Parse Chunk

kishorekrd committed:
metrics export to Victoria metrics

visualzhou committed:
Get keys. Test failed, but the skeleton is there.

visualzhou committed:
Trying to fix get keys. Getting closer.

assertion  failed
  left: 3476
 right: 3479

visualzhou committed:
Clean up code.

visualzhou committed:
Comments and minor clean up.

milkie committed:
dum

milkie committed:
improve ChunkParser plus hacks to get compression and reader to compile

milkie committed:
more code movement

milkie committed:
it compiles

milkie committed:
fixed bugs..

milkie committed:
fix more warnings

milkie committed:
cleanup

milkie committed:
use actual Metric types

visualzhou committed:
Remove unused code.

milkie committed:
fmt

visualzhou committed:
Remove compression.rs.

visualzhou committed:
Add a simple test_decode_chunk test.

milkie committed:
fix RLE

milkie committed:
got reference values working

milkie committed:
fixed the test

milkie committed:
cleanup

milkie committed:
fix timestamp increment.  add zero-count debugging.

visualzhou committed:
Polish test.

milkie committed:
timestamp?

visualzhou committed:
Use RawDocument to work around duplicated keys.

visualzhou committed:
Add test_decode_time_series.

visualzhou committed:
Add Varint reader.

visualzhou committed:
Move zero count to outmost across keys.

milkie committed:
add comments

milkie committed:
switch to VarintReader

milkie committed:
correct zero RLE logic

milkie committed:
cleanup output

milkie committed:
remove extra timestamp logic; unnecessary

milkie committed:
fix zero count

milkie committed:
cleanup debugging output

milkie committed:
fix sanitizing metric names with embedded commas.  also stop outputting all VM calls.

visualzhou committed:
Fix VM line format.

visualzhou committed:
Fix tests.

kishorekrd committed:
updated the code that clean up the old metrics data cleanup_old_metrics. added unique tags to mark the time series, like file path and folder path

visualzhou committed:
Add an list of example keys.

visualzhou committed:
Extract interface

visualzhou committed:
Move paths to the constructor.

visualzhou committed:
Extract paths and Add a new handler.

kishorekrd committed:
added --clean option to optionaly cleanup before importing

visualzhou committed:
Add test for time series reader. Fix bugs.

kishorekrd committed:
log cleanup

visualzhou committed:
Implement prometheus protobuf format.

milkie committed:
cleanup debug output and comments

kishorekrd committed:
Cleaning up metrics for the current file

visualzhou committed:
Update readme.

visualzhou committed:
Fix Clippy warnings.

visualzhou committed:
Add rust-toolchain.toml to pin toolchain version

visualzhou committed:
Add CI workflow for formatting, linting, and testing

visualzhou committed:
Add release workflow for multi-platform builds

visualzhou committed:
Fix Clippy warning: use read_to_end instead of unbuffered bytes() on ZlibDecoder

visualzhou committed:
Fix YAML syntax error in release workflow file

visualzhou committed:
Fix additional YAML syntax in release workflow permissions

visualzhou committed:
Track Cargo.lock.

visualzhou committed:
Update Cargo.toml.

visualzhou committed:
Fix musl build by adding OpenSSL dependencies

visualzhou committed:
Disable linux musl build for now.

visualzhou committed:
Add Apple Silicon (aarch64-apple-darwin) CI job

visualzhou committed:
Refactor CI to use matrix builds across platforms

visualzhou committed:
No file path or folder path. Leaving to the user.

visualzhou committed:
Add extra label.

visualzhou committed:
Change decode_time_series return type from Vec<FtdcTimeSeries> to FtdcDocumentTS

visualzhou committed:
Move timestamps from FtdcTimeSeries to FtdcDocumentTS to reduce duplication

visualzhou committed:
Implement Victoria Metrics time series import method and add test

visualzhou committed:
Keep import_document_ts as a stub for future implementation

visualzhou committed:
Remove unused imports from Victoria Metrics tests

visualzhou committed:
Use 'start' time series as document timestamps. Add tests.

visualzhou committed:
Fix timestamps.

visualzhou committed:
Escape parentheses.

milkie committed:
clean up output

visualzhou committed:
Remove unused run_import_mode_ts function

visualzhou committed:
Remove tests/ftdc_metric_parser_test.rs

visualzhou committed:
Remove .note since we didn't follow its instructions.

visualzhou committed:
Add comments.

visualzhou committed:
Remove batch size.

visualzhou committed:
Remove verify flag and related code

visualzhou committed:
Use TimeSeries in check mode.

visualzhou committed:
Fix all BSON type's conversion to i64, including i32.

visualzhou committed:
Add teset for bson to i64.

visualzhou committed:
Simplify boolean conversion and fix clingy.

visualzhou committed:
Add Linux ARM to CI platforms.

visualzhou committed:
Remove dead code of generating FTDCDocument.

visualzhou committed:
Removed unused code.

visualzhou committed:
Fix clingy warnings.

visualzhou committed:
Remove unused file_path and folder_path fields from FtdcReader

visualzhou committed:
Remove unused send_metrics method from VictoriaMetricsClient

visualzhou committed:
Remove unused MetricType enum and metric_type field

visualzhou committed:
Remove unused decode_chunk_values method and MetricValue struct

visualzhou committed:
Remove unused serde traits and imports from FtdcDocumentTS/FtdcTimeSeries

visualzhou committed:
Remove dead code.

visualzhou committed:
refactor: remove unused path parameters from ImportMetadata and use Default trait

visualzhou committed:
Only enable cache when running on GitHub Actions.

visualzhou committed:
Add native-tls-vendored instead of system SSL to avoid reduce dependency on build platform.

visualzhou committed:
Add toolchain and system SSL.

visualzhou committed:
Revert "Add native-tls-vendored instead of system SSL to avoid reduce dependency on build platform."

This reverts commit 6f69ad2da8f7987d36f04b598bf45c358a4f400d.

visualzhou committed:
Disable ARM64 build in CI due to OpenSSL cross-compilation issues

visualzhou committed:
Blank base for initial review

mpobrien commented on tests/.DS_Store:
probably don't want to commit this? maybe should be added to gitignore too.

visualzhou replied:
Removed. It's already ignored in gitignore.

mpobrien commented on tests/README.md:
```diff
@@ -0,0 +1,62 @@
+
+## Test Files
+
+- `ftdc_format_test.rs` - Detects and prints information about the different types of FTDC chunks
+- `ftdc_metric_parser_test.rs` - Analyzes the structure of a metric document
+- `ftdc_chunk_detector_test.rs` - Implements the full FTDC parsing logic
+- `ftdc_extract_metric_test.rs` - Extracts and dumps a single metric document
```
of these four, i only see `ftdc_extract_metric_test.rs` in the repo - maybe this doc is stale or the other files are pending a `git add` ?

visualzhou replied:
Sorry, this file is out of date. Will update it.

mpobrien commented on src/victoria_metrics.rs:
```diff
@@ -0,0 +1,77 @@
+impl VictoriaMetricsClient {
+    /// Create a new Victoria Metrics client
+    #[must_use]
+    pub fn new(base_url: String, metadata: ImportMetadata) -> Self {
```
It would be more idiomatic and flexible to accept `base_url: &str` instead of a `String`, and clone it here rather than require that the caller pass an owned-value.

visualzhou replied:
Done.

mpobrien commented on src/reader.rs:
```diff
@@ -0,0 +1,210 @@
+    }
+}
+
+pub type ReaderResult<T> = std::result::Result<T, FtdcError>;
```
nit: actually would be nice to just name this type `Result`. It's quite common to have package-specific result types named to shadow `Result` from the stdlib

visualzhou replied:
Done.

mpobrien commented on src/reader.rs:
```diff
@@ -0,0 +1,210 @@
+    }
+
+    /// Iterates through FTDC documents and processes them with the given handler
+    pub async fn iterate_next<H, T>(&mut self, handler: &H) -> ReaderResult<Option<T>>
```
At a glance, this looks like a use case where implementing the [Iterator trait](https://doc.rust-lang.org/std/iter/trait.Iterator.html) is worth considering here, IMO.

visualzhou replied:
That's a good point. I'll give it a try.

visualzhou replied:
I created #3 to track this idea.

mpobrien reviewed (commented):
A few quick comments. I think the maintainability/readability improvement from making this an iterator-centric API could be significant (since so much of the stdlib and rust ecosystem in general is built around the concept) but I'll leave it up to you if it's worth the squeeze!

visualzhou committed:
Remove .DS_Store files and ignore them globally

visualzhou committed:
Update tests/README.md.

visualzhou committed:
Standardize on a single Result type across the codebase

visualzhou committed:
Update Cargo.lock with recent changes

visualzhou committed:
Improve VictoriaMetricsClient API by accepting &str instead of String

visualzhou commented:
I addressed most of the comments and will try to implement the Iterator trait. @mpobrien Do you think this tool is ready to be integrated into our toolkit and the SLS script now, while we keep improving it? I'd hope to unblock the SLS team's use cases soon since there are a few use cases to view metrics from both mongod and SLS already.

mpobrien commented:
> Do you think this tool is ready to be integrated into our toolkit and the SLS script now, while we keep improving it? I'd hope to unblock the SLS team's use cases soon since there are a few use cases to view metrics from both mongod and SLS already.

Absolutely, let's do it. LGTM

visualzhou commented:
Thank you! Cherry-picked on top of master.
